p2p/net/conngater/conngater.go

```diff
@@ -0,0 +1,322 @@
+	manet "github.com/multiformats/go-multiaddr/net"
+
+	"github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-datastore/namespace"
+	"github.com/ipfs/go-datastore/query"
+	logging "github.com/ipfs/go-log"
+)
+
+type BasicConnectionGater struct {
+	sync.Mutex
+
+	blockedPeers   map[peer.ID]struct{}
+	blockedAddrs   map[string]struct{}
+	blockedSubnets map[string]*net.IPNet
+
+	ds datastore.Datastore
+}
+
+var log = logging.Logger("net/conngater")
+
+const ns = "/libp2p/net/conngater"
+const keyPeer = "/peer/"
+const keyAddr = "/addr/"
+const keySubnet = "/subnet/"
+
+// NewBasicConnectionGater creates a new connection gater.
+// The ds argument is an (optional, can be nil) datastore to persist the connection gater
+// filters
+func NewBasicConnectionGater(ds datastore.Datastore) *BasicConnectionGater {
+	cg := &BasicConnectionGater{
+		blockedPeers:   make(map[peer.ID]struct{}),
+		blockedAddrs:   make(map[string]struct{}),
+		blockedSubnets: make(map[string]*net.IPNet),
+	}
+
+	if ds != nil {
+		cg.ds = namespace.Wrap(ds, datastore.NewKey(ns))
+		cg.loadRules()
+	}
+
+	return cg
+}
+
+func (cg *BasicConnectionGater) loadRules() {
+	// load blocked peers
+	res, err := cg.ds.Query(query.Query{Prefix: keyPeer})
+	if err != nil {
+		log.Errorf("error querying datastore for blocked peers: %s", err)
+		return
+	}
+
+	for r := range res.Next() {
+		if r.Error != nil {
+			log.Errorf("query result error: %s", r.Error)
+			return
+		}
+
+		p := peer.ID(r.Entry.Value)
+		cg.blockedPeers[p] = struct{}{}
+	}
+
+	// load blocked addrs
+	res, err = cg.ds.Query(query.Query{Prefix: keyAddr})
+	if err != nil {
+		log.Errorf("error querying datastore for blocked addrs: %s", err)
+		return
+	}
+
+	for r := range res.Next() {
+		if r.Error != nil {
+			log.Errorf("query result error: %s", r.Error)
+			return
+		}
+
+		ip := net.IP(r.Entry.Value)
+		cg.blockedAddrs[ip.String()] = struct{}{}
+	}
+
+	// load blocked subnets
+	res, err = cg.ds.Query(query.Query{Prefix: keySubnet})
+	if err != nil {
+		log.Errorf("error querying datastore for blocked subnets: %s", err)
+		return
+	}
+
+	for r := range res.Next() {
+		if r.Error != nil {
+			log.Errorf("query result error: %s", r.Error)
+			return
+		}
+
+		ipnetStr := string(r.Entry.Value)
+		_, ipnet, err := net.ParseCIDR(ipnetStr)
+		if err != nil {
+			log.Errorf("error parsing CIDR subnet: %s", err)
+			return
+		}
+		cg.blockedSubnets[ipnetStr] = ipnet
+	}
+}
+
+// BlockPeer adds a peer to the set of blocked peers
+func (cg *BasicConnectionGater) BlockPeer(p peer.ID) {
+	cg.Lock()
+	defer cg.Unlock()
+
+	cg.blockedPeers[p] = struct{}{}
+
+	if cg.ds != nil {
+		err := cg.ds.Put(datastore.NewKey(keyPeer+p.String()), []byte(p))
+		if err != nil {
+			log.Errorf("error writing blocked peer to datastore: %s", err)
+		}
+	}
+}
+
+// UnblockPeer removes a peer from the set of blocked peers
+func (cg *BasicConnectionGater) UnblockPeer(p peer.ID) {
+	cg.Lock()
+	defer cg.Unlock()
+
+	delete(cg.blockedPeers, p)
```

Comment:
unlock after we're done with the thing that's guarded/protected. No reason to hold the lock during the datastore update, no?